dali/operators/imgcodec/image_decoder.h

```diff
@@ -0,0 +1,761 @@
+        if (decode_status_[i] != NVIMGCODEC_PROCESSING_STATUS_SUCCESS) {
+          int orig_idx = decode_sample_idxs_[i];
+          throw std::runtime_error(make_string("Failed to decode sample #", orig_idx, " : ",
+                                               input.GetMeta(orig_idx).GetSourceInfo()));
+        }
+      }
+    }
+
+    for (int orig_idx : decode_sample_idxs_) {
+      auto st_ptr = state_[orig_idx].get();
+      if (st_ptr->need_processing) {
+        tp_->AddWork(
+            [&, out = output[orig_idx], st_ptr, orig_idx](int tid) {
+              DomainTimeRange tr(make_string("Convert #", orig_idx), DomainTimeRange::kOrange);
+              auto &st = *st_ptr;
+              if constexpr (std::is_same<MixedBackend, Backend>::value) {
+                ConvertGPU(out, st.req_layout, st.req_img_type, st.decode_out_gpu, st.orig_layout,
+                           st.orig_img_type, ws.stream(), ROI{}, nvimgcodecOrientation_t{},
+                           st.dyn_range_multiplier);
+                st.device_buf.reset();
+              } else {
+                assert(st.dyn_range_multiplier == 1.0f);  // TODO(janton): enable
+                ConvertCPU(out, st.req_layout, st.req_img_type, st.decode_out_cpu, st.orig_layout,
+                           st.orig_img_type, ROI{}, nvimgcodecOrientation_t{});
+                st.pinned_buf.reset();
+              }
+            },
+            -orig_idx);
+      }
+    }
+    tp_->RunAll();
+
+    if (use_cache) {
+      DomainTimeRange tr(make_string("CacheStore"), DomainTimeRange::kOrange);
+      for (int orig_idx : decode_sample_idxs_) {
+        // We don't store ROIs
+        bool has_roi = rois_[orig_idx].use_roi();
+        if (has_roi)
+          continue;
+        // We only store RGB HWC data. The cache doesn't have format information, and we could have
+        // another decoder with different configuration
+        auto &st = *state_[orig_idx];
+        if (st.req_img_type != DALI_RGB || st.req_layout != "HWC")
+          continue;
+        auto src_info = input.GetMeta(orig_idx).GetSourceInfo();
+        auto *out_data = output.template mutable_tensor<uint8_t>(orig_idx);
+        const auto &out_shape = output.tensor_shape(orig_idx);
+        cache_->CacheStore(src_info, out_data, out_shape, ws.stream());
+      }
+    }
+  }
+
+  std::unique_ptr<ThreadPool> thread_pool_;
+  std::unique_ptr<CachedDecoderImpl> cache_;
+  ThreadPool *tp_ = nullptr;
+
+  NvImageCodecInstance instance_;
+  NvImageCodecDecoder decoder_;
+
+  nvimgcodecExecutorDesc_t executor_{NVIMGCODEC_STRUCTURE_TYPE_EXECUTOR_DESC,
+                                     sizeof(nvimgcodecExecutorDesc_t),
+                                     nullptr,
+                                     this,
+                                     &static_launch,
+                                     &static_get_num_threads};
+  nvimgcodecDeviceAllocator_t dev_alloc_;
+  nvimgcodecPinnedAllocator_t pinned_alloc_;
```

Comment:
```suggestion
  nvimgcodecDeviceAllocator_t dev_alloc_ = {};
  nvimgcodecPinnedAllocator_t pinned_alloc_ = {};
```
Something exactly like that has already bitten us in optical flow.